ADDING_A_PLUGIN.md

```diff
@@ -0,0 +1,347 @@
+  - [x] Also - provide a non TF way to write plugin data. Describe API of the `plugins/$foo/summary.py`, which exposes `op` (TF) and `pb` (non-TF).
+  - [x] *(nothing to say)* Tips & Tricks for consistent design with existing plugins
+  - [x] Include a demo script that generates representative data that exercises all the features of your plugin, preferably following the API of existing such scripts.
+  -->
+
+### Backend: How the plugin processes data, and sends it to the browser
+
+> **Note:** The backend APIs are under active development and might change. If you [create a GitHub issue] telling us that you're working on a plugin, then we can proactively reach out and help you update your code when API changes occur.
+
+[create a GitHub issue]: https://github.com/tensorflow/tensorboard/issues/new
+
+To provide routes (endpoints) for your plugin, create a subclass of `BasePlugin`. Your subclass must have a static `plugin_name` property, and you'll need to implement the following methods:
+
+  - `__init__`: Your plugin will be passed a [`TBContext`] when it is constructed. Among other things, this context exposes an *[event multiplexer]*, which is used to access the actual summary data. You'll probably want to store this onto your plugin object for later use.
+  - `is_active`: This should return whether the plugin is active (whether there exists relevant data for the plugin to process). TensorBoard will exclude inactive plugins from the main navigation bar.
+  - `get_plugin_apps`: This should return a dictionary mapping routes to WSGI applications. More on this below.
+
+[`TBContext`]: https://github.com/tensorflow/tensorboard/blob/master/tensorboard/plugins/base_plugin.py#L72
+[event multiplexer]: https://github.com/tensorflow/tensorboard/blob/master/tensorboard/backend/event_processing/event_multiplexer.py
+
+To retrieve relevant data (tensors) read from summaries, the logic of your plugin's route should call `PluginRunToTagToContent` with the `plugin_name`. This returns a dictionary mapping from run name to all of the tags that are associated with your plugin. The tag names themselves map to the `content` from the `PluginData` proto.
+
+Subsequently, with knowledge of which tags are relevant, your plugin can retrieve the tensors relevant to each tag. Call the `Tensors` method of the `Multiplexer` object to retrieve the tensors sampled by TensorBoard.
+
+Below, `GreeterPlugin` demonstrates how a very basic plugin works. To peruse the code of a real plugin, see [`TextPlugin`](https://github.com/tensorflow/tensorboard/blob/efd5909b2ff60c2579b6052ed3c0e67402235623/tensorboard/plugins/text/text_plugin.py) or any of the plugins within TensorBoard's GitHub repository.
+
+```python
+class GreeterPlugin(base_plugin.TBPlugin):
+  """A plugin that serves greetings recorded during model runs."""
+
+  # This static property will also be included within routes (URL paths)
+  # offered by this plugin, so it must be URL-friendly. This property
```

Comment:
I just removed the "must be url friendly" bit. If they do the wrong thing, they'll find out quickly.